src/transport/raw/tests/NetworkTestHelpers.h

```diff
@@ -61,35 +61,30 @@ class IOContext
     Inet::EndPointManager<Inet::UDPEndPoint> * mUDPEndPointManager = nullptr;
 };
 
-class LoopbackTransport : public Transport::Base
+class LoopbackTransportDelegate
 {
 public:
-    /// Transports are required to have a constructor that takes exactly one argument
-    CHIP_ERROR Init(const char *) { return CHIP_NO_ERROR; }
+    virtual ~LoopbackTransportDelegate() {}
+    virtual void OnMessageDropped() {}
+};
 
-    /*
-     * For unit-tests that simulate end-to-end transmission and reception of messages in loopback mode,
-     * this mode better replicates a real-functioning stack that correctly handles the processing
-     * of a transmitted message as an asynchronous, bottom half handler dispatched after the current execution context has
-     * completed. This is achieved using SystemLayer::ScheduleWork.
-     */
-    void EnableAsyncDispatch(System::Layer * aSystemLayer)
+class LoopbackTransport : public Transport::Base
+{
+public:
+    void InitLoopbackTransport(System::Layer * systemLayer) { mSystemLayer = systemLayer; }
+    void ShutdownLoopbackTransport()
     {
-        mSystemLayer          = aSystemLayer;
-        mAsyncMessageDispatch = true;
+        while (!mPendingMessageQueue.empty())
+            mPendingMessageQueue.pop();
```

Comment:
Why do we need to do this by hand instead of just waiting for our destructor to clear it?  At the very least, needs to be documented.

Reply:
We can't, the destructor runs after `Platform::MemoryShutdown` being called, resulting an abort. I'll add comments for it

Reply:
Thank you.  Why did we not need this before?  Were all the tests getting through the whole queue but now we have some that are not?

Reply:
@kghost and I prodded and this and discovered that:
1. TestReportingEngine's `TestBuildAndSendSingleReportData` is in fact missing a `DrainAndServiceIO` call.
2. This is being covered up right now by destructor ordering: the loopback transport is destroyed before the MessagingContext, and that has a superclass whose destructor does the final MemoryShutdown.
3. With the changes in this PR, the destructor of LoopbackTransportManager runs after the MessagingContext destructor, so now we hit the attempts to deallocate our messages after final MemoryShutdown.

As a followup, the plan is to fix the test and make LoopbackTransportManager shutdown assert that there are no pending messages hanging around.